### PR TITLE
Alerting: Add support for tracing to alerting scheduler

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR adds support for tracing to alerting the scheduler. 
The name of the span: `alert rule execution`
The span attributes:
- `rule_uid` - UID of the rule
- `org_id` - ID of the organization. (for consistency, I found that other spans use it. We can drop it)
- `rule_version` - version of the rule
- `tick` - the formatted timestamp of the tick.

Span events:
- After evaluation of the rule's query. If error an event with the message `rule evaluation failed` and errors. If normal - `rule evaluated` and number of results (dimensions)
- `results processed` after state processing with a number of state transitions returned by the state manager and a number of alerts to send to the notifier.

<details><summary>Example</summary>

image

</details>

**Why do we need this feature?**
This will help get more insights into how the background service behaves

**Which issue(s) does this PR fix?**:
Related: https://github.com/grafana/grafana/issues/55196

### CODE REVIEW

File: pkg/services/ngalert/schedule/schedule.go [hunk 1]
```diff
@@ -363,15 +370,42 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
...
 				results = append(results, eval.NewResultFromError(err, e.scheduledAt, dur))
 			}
+			if err == nil {
+				for _, result := range results {
+					if result.Error != nil {
+						err = multierror.Append(err, result.Error)
+					}
+				}
+			}
+			span.RecordError(err)
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/schedule/schedule.go:380]: If we pass `nil` to `RecordError` is it a no-op?

File: pkg/services/ngalert/schedule/schedule.go [hunk 2]
```diff
@@ -363,15 +370,42 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
...
+			span.AddEvents(
+				[]string{"error", "message"},
+				[]tracing.EventValue{
+					{Str: fmt.Sprintf("%v", err)},
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/schedule/schedule.go:384]: Will this be `"nil"` when `err` is nil? Is that what we want? If it is, please feel free to close this comment!

File: pkg/services/ngalert/schedule/schedule.go [hunk 3]
```diff
@@ -363,15 +370,42 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
...
+					{Str: "rule evaluation failed"},
+				})
 		} else {
 			logger.Debug("Alert rule evaluated", "results", results, "duration", dur)
+			span.AddEvents(
+				[]string{"message", "results"},
+				[]tracing.EventValue{
+					{Str: "rule evaluated"},
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/schedule/schedule.go:392]: Is this the number of rules evaluated?

```suggestion
					{Str: "rule(s) evaluated"},
```

File: pkg/services/ngalert/schedule/schedule.go [hunk 4]
```diff
@@ -434,7 +468,14 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
...
 						currentRuleVersion = newVersion
 					}
-					evaluate(grafanaCtx, attempt, ctx)
+					tracingCtx, span := sch.tracer.Start(grafanaCtx, "alert rule execution")
+					span.SetAttributes("rule_uid", ctx.rule.UID, attribute.String("rule_uid", ctx.rule.UID))
+					span.SetAttributes("org_id", ctx.rule.OrgID, attribute.Int64("org_id", ctx.rule.OrgID))
+					span.SetAttributes("rule_version", ctx.rule.Version, attribute.Int64("rule_version", ctx.rule.Version))
+					utcTick := ctx.scheduledAt.UTC().Format(time.RFC3339Nano)
+					span.SetAttributes("tick", utcTick, attribute.String("tick", utcTick))
+					defer span.End()
```
Comment (grobinson-grafana) [inline pkg/services/ngalert/schedule/schedule.go:477]: Can we put the `defer` on Line 472?

**Review by grobinson-grafana — APPROVED** [review]

File: pkg/services/ngalert/schedule/schedule.go (hunk 1 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:380]: nil is not possible. That branch is executed when either `err` is not nil or `results` have errors, i.e. there is at least one result that contains error and therefore `err` will be assigned.

File: pkg/services/ngalert/schedule/schedule.go (hunk 3 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:392]: no, this event of a single span. Kind of checkpoint to measure how long it took to evaluate a rule

File: pkg/services/ngalert/schedule/schedule.go (hunk 2 above)
Comment (yuri-tceretian) [reply pkg/services/ngalert/schedule/schedule.go:384]: nil is not possible here

**Review by grobinson-grafana — APPROVED** [review]

**Comment by frodgim** [comment]
Hi all! 
How could I open this view in my grafana 9.4.0-beta? I don't see the option "alert rule execution" in any alert instance raised.
Thank you in advance!

**Comment by yuri-tceretian** [comment]
Hi @frodgim tracing is not alerting feature. To be able to see traces you need run Tempo or Jaeger (applications for collecting traces) and configure your Grafana instance to send traces there. 
Please refer to https://github.com/grafana/grafana/blob/main/contribute/backend/instrumentation.md#how-to-collect-visualize-and-query-traces-and-correlate-logs-with-traces-locally for how to do it in a local dev environment

**Pull request merged**